go_lib/hooks/tls_certificate/internal_tls.go

```diff
@@ -0,0 +1,267 @@
+					MatchNames: []string{conf.TLSSecretName},
+				},
+
+				FilterFunc: tlsFilter,
+			},
+		},
+	}, genSelfSignedTLS(conf))
+}
+
+func tlsFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var secret v1.Secret
+	err := sdk.FromUnstructured(obj, &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return certificate.Certificate{
+		CA:   string(secret.Data["ca.crt"]),
+		Cert: string(secret.Data["tls.crt"]),
+		Key:  string(secret.Data["tls.key"]),
+	}, nil
+}
+
+func genSelfSignedTLS(conf GenSelfSignedTLSHookConf) func(input *go_hook.HookInput) error {
+	caPath, certPath, keyPath := conf.generatePaths()
+
+	return func(input *go_hook.HookInput) error {
+		var cert certificate.Certificate
+		var err error
+
+		cn, sans := conf.CN, conf.SANs(input)
+
+		if len(input.Snapshots["secret"]) == 0 {
+			// No certificate in snapshot => generate a new one.
+			// Secret will be updated by Helm.
+			cert, err = generateNewTLS(input, cn, sans)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Certificate is in the snapshot => load it.
+			cert = input.Snapshots["secret"][0].(certificate.Certificate)
+			// update certificate if less than 6 month left. We create certificate for 10 years, so it looks acceptable
+			// and we don't need to create Crontab schedule
+			caOutdated, err := isOutdatedCA(cert.CA)
+			if err != nil {
+				return err
+			}
+			certOutdated, err := isOutdatedCert(cert.Cert, sans)
+			if err != nil {
+				return err
+			}
+
+			if caOutdated || certOutdated {
+				cert, err = generateNewTLS(input, cn, sans)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Note that []byte values will be encoded in base64. Use strings here!
+		input.Values.Set(caPath, cert.CA)
+		input.Values.Set(certPath, cert.Cert)
+		input.Values.Set(keyPath, cert.Key)
+		return nil
+	}
+}
+
+func isOutdatedCert(certData string, desiredSANSs []string) (bool, error) {
+	cert, err := certificate.ParseCertificate(certData)
+	if err != nil {
+		return false, err
+	}
+
+	if time.Until(cert.NotAfter) < certOutdateDuration {
+		return true, nil
+	}
+	var dnsNames, ipAddrs []string
+	for _, san := range desiredSANSs {
+		if net.IsIPv4String(san) {
+			ipAddrs = append(ipAddrs, san)
+		} else {
+			dnsNames = append(dnsNames, san)
+		}
+	}
+
+	if !arrayAreEqual(dnsNames, cert.DNSNames) {
+		return true, nil
+	}
+
+	if len(ipAddrs) > 0 {
+		certIPs := make([]string, 0, len(cert.IPAddresses))
+		for _, cip := range cert.IPAddresses {
+			certIPs = append(certIPs, cip.String())
+		}
+		if !arrayAreEqual(ipAddrs, certIPs) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func isOutdatedCA(ca string) (bool, error) {
+	cert, err := certificate.ParseCertificate(ca)
+	if err != nil {
+		return false, err
+	}
+
+	if time.Until(cert.NotAfter) < certOutdateDuration {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func generateNewTLS(input *go_hook.HookInput, cn string, sans []string) (certificate.Certificate, error) {
```

Comment:
```suggestion
func generateNewSelfSignedTLS(input *go_hook.HookInput, cn string, sans []string) (certificate.Certificate, error) {
```
Because it now lies in the generic tls certificate library (but there is already a function with this name :thinking:).